File: php/79-word-search.php
<?php 
class Solution {
    private $direction = [
        [0, -1],
        [1, 0],
        [0, 1],
        [-1, 0]
    ];
    private $col;
    private $row;
    private $visited = [];
    /**
     * @param String[][] $board
     * @param String $word
     * @return Boolean
     */
    function exist($board, $word) {
        $this->row = count($board);
        $this->col = count($board[0]);
        $this->visited = array_fill(0, $this->row, array_fill(0, $this->col, false));
        for ($i = 0; $i < $this->row; $i++) {
            for ($j = 0; $j < $this->col; $j++) {
                $res = $this->getWord($board, $word, 0, $i, $j);
                if ($res) {
                    return true;
                }
            }
        }
        return false;
    }
    function getWord(&$board, $word, $index, $x, $y) {
        if ($index == strlen($word) - 1) {
            return $board[$x][$y] == $word[$index];
        }
        if ($word[$index] !== $board[$x][$y]) {
            return false;
        }
        $this->visited[$x][$y] = true;
        // continure search
        for ($i = 0; $i < 4; $i ++) {
            $nextX = $x + $this->direction[$i][0];
            $nextY = $y + $this->direction[$i][1];
            $isVisited = isset($this->visited[$nextX][$nextY]) && $this->visited[$nextX][$nextY];
            if ($this->inArea($nextX, $nextY) && !$isVisited) {
                if ($this->getWord($board, $word, $index+1, $nextX, $nextY)) {
                    return true;                    
                }
            }
        }
        $this->visited[$x][$y] = false;
        return false;
    }
    function inArea($x, $y) {
        return $x >=0 && $x < $this->row && $y >= 0 && $y < $this->col;
    }
}

File: go/79-word-search.go
package main

var direction = [][]int{
	{0, -1},
	{1, 0},
	{0, 1},
	{-1, 0},
}
var visited [][]bool
var row, col int

func exist(board [][]byte, word string) bool {
	row = len(board)
	col = len(board[0])
	visited = make([][]bool, row)
	//init visited
	for i := 0; i < row; i++ {
		visited[i] = make([]bool, col)
	}
	for j := 0; j < row; j++ {
		for k := 0; k < col; k++ {
			if wordExist(board, word, j, k, 0) {
				return true
			}
		}
	}
	return false
}

func wordExist(board [][]byte, word string, x, y, index int) bool {
	c := byte(word[index : index+1][0])
	if index == len(word)-1 {
		return c == board[x][y]
	}
	if c != board[x][y] {
		return false
	}
	// four direction search
	visited[x][y] = true
	for d := 0; d < 4; d++ {
		nextX := x + direction[d][0]
		nextY := y + direction[d][1]
		if inBoard(nextX, nextY) && !visited[nextX][nextY] {
			if wordExist(board, word, nextX, nextY, index+1) {
				return true
			}
		}
	}
	visited[x][y] = false
	return false
}

func inBoard(x, y int) bool {
	if x < 0 || y < 0 || x > row-1 || y > col-1 {
		return false
	}
	return true
}
